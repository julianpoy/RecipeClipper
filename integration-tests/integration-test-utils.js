import "regenerator-runtime/runtime";
import * as jsdom from "jsdom";
import sanitizeHtml from "sanitize-html";

import { clipRecipe } from '../dist/recipe-clipper.esm';

const replaceBrWithBreak = (html) => {
  return html.replaceAll(new RegExp(/<br( \/)?>/, "g"), "\n");
};

export const clipRecipeFromUrl = async clipUrl => {
  const response = await fetch(clipUrl);
  const html = await response.text();

  const dom = new jsdom.JSDOM(html);
  const { window } = dom;

  Object.defineProperty(window.Element.prototype, 'innerText', {
    get() {
      const html = replaceBrWithBreak(this.innerHTML);
      return sanitizeHtml(html, {
        allowedTags: [], // remove all tags and return text content only
        allowedAttributes: {}, // remove all tags and return text content only
      });
    },
  });

  window.fetch = fetch;

  return await clipRecipe({
    window,
    mlClassifyEndpoint: 'http://localhost:3060',
  });
};

