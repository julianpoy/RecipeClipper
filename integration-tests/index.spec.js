import { clipRecipeFromUrl } from './integration-test-utils'

const sites = [
  'https://www.bettycrocker.com/recipes/ultimate-chocolate-chip-cookies/77c14e03-d8b0-4844-846d-f19304f61c57'
];

describe('testing', () => {
  sites.forEach(site => {
    it(`clips recipe matching snapshot from ${site}`, async () => {
      const clip = await clipRecipeFromUrl(site)
      expect(clip).toMatchSnapshot();
    })
  })
})
