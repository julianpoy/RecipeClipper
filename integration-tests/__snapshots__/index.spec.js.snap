// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing clips recipe matching snapshot from https://food52.com/recipes/81867-best-quiche-recipe 1`] = `
Object {
  "activeTime": "3 Hours",
  "description": "",
  "imageURL": "https://images.food52.com/LVMXJEEPYuxKMtj9wsyPRqnMXrY=/965x643/filters:format(webp)/ad25bacb-a4f9-406e-af6b-4f48fc34ed3a--2019-0820_best-quiche-recipe_3x2_bobbi-lin_083.jpg",
  "ingredients": "Ingredients send grocery list
Crust
1 cup (128 grams) all-purpose flour
1/3 cup (40 grams) whole-wheat flour
1/2 teaspoon kosher salt
1/2 teaspoon granulated sugar
1/8 teaspoon baking powder
3 ounces cream cheese, very cold, cubed
4 ounces (1 stick) unsalted butter, very cold, cubed
1 1/2 tablespoons water, very cold
1 1/2 teaspoons apple cider vinegar
1 large egg white
Filling
1 cup heavy cream
1 cup whole milk
3 tablespoons all-purpose flour
5 large eggs
1 large egg yolk
1 3/4 teaspoons kosher salt
3/4 teaspoon freshly ground black pepper
1 to 1 1/4 cups mix-ins, depending on mix-in (see headnote)
1 cup grated cheese (see headnote)",
  "instructions": "",
  "notes": "",
  "source": "How to Make Any Type of Quiche Perfectly",
  "title": "Best Quiche Recipe",
  "totalTime": "",
  "yield": "",
}
`;

exports[`testing clips recipe matching snapshot from https://myheartbeets.com/paleo-tortilla-chips/ 1`] = `
Object {
  "activeTime": "",
  "description": "",
  "imageURL": "https://myheartbeets.com/wp-content/uploads/2015/08/paleo-tortilla-chips.jpg",
  "ingredients": "1 cup coconut milk, full-fat
½ cup almond flour
½ cup tapioca flour
½ teaspoon salt, adjust to taste
¼ teaspoon turmeric",
  "instructions": "Pre-heat a 9.5 inch (or larger) nonstick pan over medium heat.
Mix the almond flour, tapioca flour, coconut milk, salt and turmeric together in a bowl.
Add ¼ cup worth of batter onto the pan.
After about a minute flip the flatbread and cook for another minute or until both sides are mostly firm, then place the bread onto a parchment-lined baking sheet. Repeat until the batter is done.
Using scissors, cut the tortillas into small triangles and spread them out on the baking sheet.
Bake at 400°F for 10-12 minutes or until the tortilla chips are crispy.",
  "notes": "",
  "source": "My Heart Beets",
  "title": "Paleo Tortilla Chips",
  "totalTime": "",
  "yield": "",
}
`;

exports[`testing clips recipe matching snapshot from https://panlasangpinoy.com/leche-flan/ 1`] = `
Object {
  "activeTime": "10 Minutes",
  "description": "This is a simple and easy version of a famous Filipino dessert known as Leche Flan.",
  "imageURL": "https://panlasangpinoy.com/wp-content/uploads/2009/07/Leche-Flan-Recipe-2.jpg",
  "ingredients": "10 pieces eggs
1 can condensed milk (14 oz)
1 cup fresh milk or evaporated milk
1 cup granulated sugar
1 teaspoon vanilla extract
US Customary - Metric",
  "instructions": "Using all the eggs, separate the yolk from the egg white (only egg yolks will be used).
Place the egg yolks in a big bowl then beat them using a fork or an egg beater
Add the condensed milk and mix thoroughly
Pour-in the fresh milk and Vanilla. Mix well
Put the mold (llanera) on top of the stove and heat using low fire
Put-in the granulated sugar on the mold and mix thoroughly until the solid sugar turns into liquid (caramel) having a light brown color. Note: Sometimes it is hard to find a Llanera (Traditional flan mold) depending on your location. I find it more convenient to use individual Round Pans in making leche flan.
Spread the caramel (liquid sugar) evenly on the flat side of the mold
Wait for 5 minutes then pour the egg yolk and milk mixture on the mold
Cover the top of the mold using an Aluminum foil
Steam the mold with egg and milk mixture for 30 to 35 minutes.
After steaming, let the temperature cool down then refrigerate
Serve for dessert. Share and Enjoy!",
  "notes": "",
  "source": "panlasangpinoy.com",
  "title": "Leche Flan",
  "totalTime": "55 Minutes",
  "yield": "6 People",
}
`;

exports[`testing clips recipe matching snapshot from https://sallysbakingaddiction.com/quiche-recipe/ 1`] = `
Object {
  "activeTime": "2 Hours, 40 Minutes",
  "description": "",
  "imageURL": "https://cdn.sallysbakingaddiction.com/wp-content/uploads/2019/04/quiche-600x900.jpg",
  "ingredients": "1 unbaked Flaky Pie Crust (what I used) or All Butter Pie Crust*
4 large eggs
1/2 cup whole milk*
1/2 cup heavy cream or heavy whipping cream*
1/4 teaspoon each salt and pepper*
up to 1 cup shredded or crumbled cheese (see recipe note)
up to 2 cups add-ins (see recipe note)
optional toppings for serving: extra cheese, chopped herbs, hollandaise sauce, & freshly ground pepper to taste",
  "instructions": "Pie crust: I like to make sure my pie dough is prepared before I begin the quiche. Make pie dough the night before because it needs to chill in the refrigerator for at least 2 hours before rolling out and blind baking (next step).
Preheat oven to 400°F (204°C). Partially blind bake your pie crust. Follow blind baking instructions through step 9 in your 9-inch pie dish. Step 4 is optional, though that step guarantees thick pie crust edges. Crust can still be warm when you pour in the filling. (You can partially pre-bake the crust up to 3 days ahead of time. Cover cooled crust tightly and refrigerate until ready to fill.)
Reduce oven temperature to 350°F (177°C).
In a large bowl with a handheld or stand mixer fitted with a whisk attachment, beat the eggs, whole milk, heavy cream, salt, and pepper together on high speed until completely combined, about 1 minute. Whisk in add-ins.
Bake the quiche until the center is just about set, about 45-55 minutes. Don’t over-bake. Use a pie crust shield to prevent the pie crust edges from over-browning. Allow to cool for 15 minutes. Top with optional toppings before slicing and serving, if desired. Or you can cool the quiche completely before serving– it’s fantastic at room temperature!
This quiche makes great leftovers! Cover tightly and store in the refrigerator for up to 4 days.",
  "notes": "",
  "source": "Sally's Baking Addiction",
  "title": "Perfect Quiche Recipe (any Flavor!)",
  "totalTime": "4 Hours",
  "yield": "Serves 8",
}
`;

exports[`testing clips recipe matching snapshot from https://www.allrecipes.com/recipe/231244/asparagus-mushroom-bacon-crustless-quiche/ 1`] = `
Object {
  "activeTime": "",
  "description": "",
  "imageURL": "https://imagesvc.meredithcorp.io/v3/mm/image?url=https%3A%2F%2Fimages.media-allrecipes.com%2Fuserphotos%2F971958.jpg&w=640&h=640&c=sc&poi=face&q=85",
  "ingredients": "Original recipe yields 8 servings
Ingredient Checklist
6 ounces bacon, diced
½ pound asparagus, cut into 1-inch pieces
4 ounces baby bella (cremini) mushrooms, sliced
4 ounces shredded Swiss cheese
1 ½ cups heavy whipping cream
4 large eggs
¼ teaspoon salt
freshly ground black pepper to taste",
  "instructions": "Instructions Checklist
Position oven rack to center position. Preheat oven to 400 degrees F (200 degrees C). Butter a 10-inch pie plate.
Heat a large skillet over medium heat; cook and stir bacon in the hot skillet until crisp, 5 to 10 minutes. Remove bacon and drain all but 2 tablespoons grease from skillet. Cook and stir asparagus and mushrooms in the remaining bacon grease until asparagus is tender, about 5 minutes.
Remove skillet from heat and toss bacon into asparagus mixture. Spread asparagus-bacon mixture into the prepared pie pan. Sprinkle Swiss cheese over mixture.
Whisk cream, eggs, salt, and pepper in a bowl. Pour egg mixture over asparagus-cheese mixture.
Bake in the preheated oven until quiche is set and a knife inserted in the center comes out clean, about 30 minutes.
I Made It  Print",
  "notes": "Partner Tip
Reynolds® Aluminum foil can be used to keep food moist, cook it evenly, and make clean-up easier.",
  "source": "Allrecipes.com",
  "title": "Asparagus Mushroom Bacon Crustless Quiche Recipe",
  "totalTime": "",
  "yield": "",
}
`;

exports[`testing clips recipe matching snapshot from https://www.bettycrocker.com/recipes/ultimate-chocolate-chip-cookies/77c14e03-d8b0-4844-846d-f19304f61c57 1`] = `
Object {
  "activeTime": "",
  "description": "",
  "imageURL": "https://images-gmi-pmc.edge-generalmills.com/087d17eb-500e-4b26-abd1-4f9ffa96a2c6.jpg",
  "ingredients": "2 1/4
cups Gold Medal™ all-purpose flour
teaspoon baking soda
1/2
teaspoon salt
cup butter, softened
3/4
cup granulated sugar
3/4
cup packed brown sugar
egg
teaspoon vanilla
cups semisweet chocolate chips
cup coarsely chopped nuts, if desired",
  "instructions": "",
  "notes": "",
  "source": "BettyCrocker.com",
  "title": "Ultimate Chocolate Chip Cookies Recipe",
  "totalTime": "",
  "yield": "48",
}
`;

exports[`testing clips recipe matching snapshot from https://www.bonappetit.com/recipe/chocolate-babka 1`] = `
Object {
  "activeTime": "",
  "description": "",
  "imageURL": "https://assets.bonappetit.com/photos/57accbaa1b334044149751c1/1:1/w_768,h_768,c_limit/chocolate-babka1.jpg",
  "ingredients": "Dough
½ cup whole milk, warmed
1 ¼-ounce envelope active dry yeast (about 2¼ teaspoons)
¼ cup granulated sugar, plus more
1 large egg
1 large egg yolk
½ teaspoon kosher salt
2 cups all-purpose flour, plus more for surface
7 tablespoons unsalted butter, room temperature, cut into pieces, plus more for bowl
Filling and Assembly
4 tablespoons unsalted butter, melted, divided, plus more room temperature for pan
3 ounces bittersweet chocolate, finely chopped
2 tablespoons unsweetened cocoa powder
2 tablespoons granulated sugar, divided
¾ teaspoon ground cinnamon, divided
¼ teaspoon kosher salt, divided
2 tablespoons dark brown sugar
5 tablespoons all-purpose flour
1 large egg
1 tablespoon heavy cream",
  "instructions": "[Recipe Preparation]
Dough
Pour milk into a measuring glass or small bowl; sprinkle yeast and a pinch of granulated sugar over milk. Let stand until foamy, about 5 minutes.
Whisk egg, egg yolk, and remaining ¼ cup granulated sugar in a medium bowl until smooth. Whisk in yeast mixture.
Combine salt and 2 cups flour in the bowl of an electric mixer fitted with the paddle attachment. Add egg mixture and beat on low speed until nearly incorporated, about 30 seconds. Switch to dough hook and add 7 Tbsp. butter; beat on medium-low until butter is completely incorporated and dough is smooth and just slightly sticky, 8–10 minutes.
Turn out dough onto a lightly floured surface and knead a few times until very smooth (dough will still be wet and a little sticky). Transfer to a large buttered bowl. Cover with plastic wrap and let rise in a warm place until doubled in size, 1½–2 hours, then chill 45 minutes.
Filling and Assembly
Preheat oven to 350°. Generously butter a 8½x4½\\" loaf pan. Heat chocolate, cocoa, 2 Tbsp. butter, 1 Tbsp. granulated sugar, ¼ tsp. cinnamon, and ⅛ tsp. salt in a heatproof bowl set over a saucepan of simmering water, stirring occasionally, until chocolate and butter are melted and mixture is smooth. (Alternatively, microwave on medium power 1 minute, stirring halfway through.)
Turn out chilled dough onto a clean lightly floured surface. Roll out to a 22x12\\" rectangle and orient so a long side is facing you. Using an offset spatula, spread chocolate filling over dough to extend to the edges. Roll up dough away from you like a jelly roll, pulling lightly on it as you roll to maintain thickness. (Use a metal bench scraper if needed to help dough release from surface.)
Cut log in half crosswise. Set halves so they are side by side and touching. A long side should still be facing you. Place one half over the other to make an X, then twist the two ends on one side twice. Repeat on opposite side. (You should have a total of 4 twists.) Transfer bread to pan, cover with plastic wrap, and let rise 30 minutes.
Meanwhile, make the streusel. Whisk brown sugar and remaining 1 Tbsp. granulated sugar, ½ tsp. cinnamon, and ⅛ tsp. salt in medium bowl to blend. Mix in remaining 2 Tbsp. butter. Add flour and toss with a fork until moist clumps form (streusel will look slightly wet). Whisk egg and cream in a small bowl and brush egg wash over loaf. Scatter streusel over top.
Bake babka, rotating pan halfway through, until golden brown, 50–60 minutes. Transfer pan to a wire rack and let babka cool in pan 15 minutes, then turn out onto rack, running a paring knife around edges to help loosen if needed. Let cool completely before slicing.",
  "notes": "",
  "source": "Bon Appetit",
  "title": "Chocolate Babka Recipe",
  "totalTime": "",
  "yield": "",
}
`;

exports[`testing clips recipe matching snapshot from https://www.delish.com/cooking/recipe-ideas/a25648042/crustless-quiche-recipe/ 1`] = `
Object {
  "activeTime": "0 Hours 20 Mins",
  "description": "",
  "imageURL": "https://hips.hearstapps.com/hmg-prod.s3.amazonaws.com/images/crustless-quiche-128-1546439725.jpg?crop=0.668xw:1.00xh;0.175xw,0&resize=480:*",
  "ingredients": "1 tbsp.
butter
8 oz.
cremini mushrooms, thinly sliced
shallot, minced
2 c.
loosely packed spinach
Kosher salt
Freshly ground black pepper
large eggs
1/4 c.
whole milk
1/4 c.
oil-packed sun-dried tomatoes, finely chopped
1/4 c.
freshly grated Parmesan",
  "instructions": "Preheat oven to 375°. In a medium skillet over medium heat, melt butter. Add mushrooms and let cook, undisturbed, for 2 minutes. Stir and continue to cook until mushrooms are tender and golden, 5 to 6 minutes. Add shallot and cook until fragrant, 1 minute. Add spinach and cook until wilted, 1 minute more. Season with salt and pepper and remove from heat.
In a large bowl, whisk together eggs, milk, tomatoes, and Parmesan. Fold in the mushroom mixture and season again with salt and pepper. Pour into a 8\\" to 9\\" pie dish and bake until eggs are just set, 18 to 20 minutes. Let cool 3 minutes before slicing and serving.",
  "notes": "",
  "source": "How To Make Crustless Quiche",
  "title": "Best Crustless Quiche Recipe",
  "totalTime": "0 Hours 40 Mins",
  "yield": "[s]
6 Servings",
}
`;

exports[`testing clips recipe matching snapshot from https://www.foodnetwork.com/recipes/paula-deen/spinach-and-bacon-quiche-recipe-2131172 1`] = `
Object {
  "activeTime": "",
  "description": "",
  "imageURL": "https://food.fnr.sndimg.com/content/dam/images/food/fullset/2011/4/7/0/PA0501_spinach-and-bacon-quiche_s4x3.jpg.rend.hgtvcom.826.620.suffix/1371595433466.jpeg",
  "ingredients": "6 large eggs, beaten
1 1/2 cups heavy cream
Salt and pepper
2 cups chopped fresh baby spinach, packed
1 pound bacon, cooked and crumbled
1 1/2 cups shredded Swiss cheese
1 (9-inch) refrigerated pie crust, fitted to a 9-inch glass pie plate",
  "instructions": "Preheat the oven to 375 degrees F.
Combine the eggs, cream, salt, and pepper in a food processor or blender. Layer the spinach, bacon, and cheese in the bottom of the pie crust, then pour the egg mixture on top. Bake for 35 to 45 minutes until the egg mixture is set. Cut into 8 wedges.",
  "notes": "",
  "source": "Paula Deen",
  "title": "Spinach And Bacon Quiche Recipe",
  "totalTime": "",
  "yield": "8 Servings",
}
`;

exports[`testing clips recipe matching snapshot from https://www.kingarthurflour.com/recipes/chocolate-cake-recipe 1`] = `
Object {
  "activeTime": "",
  "description": "",
  "imageURL": "https://www.kingarthurflour.com/sites/default/files/styles/featured_image/public/recipe_legacy/47-3-large.jpg?itok=CfPiwrEc",
  "ingredients": "2 1/4 cups (269g) King Arthur Unbleached All-Purpose Flour
1 1/2 teaspoons baking powder
1/2 teaspoon baking soda
1/2 teaspoon salt
3/4 cup (64g) unsweetened cocoa, natural or Dutch-process cocoa
1 3/4 cups (354g) granulated sugar
1/2 cup (8 tablespoons, 113g) unsalted butter, at room temperature
1/3 cup (64g) vegetable oil
2 teaspoons vanilla extract
1 cup (227g) milk
1/2 cup (113g) coffee or water
4 large eggs
[Nutrition Information]",
  "instructions": "Preheat the oven to 350°F. Lightly grease and flour your choice of pans: one 9\\" x 13\\" pan, two 9\\" round pans, or three 8\\" round pans.
Sift together the flour, baking powder, baking soda, salt, cocoa, and sugar into a large mixing bowl.
Add the butter and mix at low speed for 1 minute. With the mixer running, add the oil and continue mixing until the mixture looks sandy, about 30 seconds more.
In a large measuring cup, combine the vanilla with the milk and coffee(or water), and add all at once to the dry ingredients. Mix for 1 minute at low speed, stop and scrape the sides and bottom of the bowl, then mix for 30 seconds more.
Take it a step further
[Blog]
How to tell when cake is done
[By Kye Ameden]
Add the eggs one at a time, beating well at medium-high speed between additions.
Scrape the sides and bottom of the mixing bowl, and mix for 30 seconds more. The batter will be thin.
Transfer the batter to the prepared pan(s) of your choice. Smooth out the tops of the layers with an offset spatula.
Bake for 34 minutes for a 9\\" x 13\\" pan, 28 to 30 minutes for 9\\" layers, and 24 to 26 minutes for 8\\" layers.
The cake is done when the top springs back when very lightly touched in the center, and the edges just begin to pull away from the edge of the pan. A cake tester inserted in the center will come out clean.
Remove the cake from the oven and let cool completely in the pan(s) on a rack before turning out of the pans to frost.
Cake layers can be wrapped and frozen for up to a month before frosting.
Tips from our Bakers
The volume of this batter is 5 3/4 cups. It will rise quite a bit during baking. The weight of the batter is 3 pounds, 1 ounce. If you're weighing out layers, put 1 pound, 8 1/2 ounces in each 9\\" layer, or 1 pound in each 8\\" layer.
We've tested this cake with natural unsweetened cocoa, and with Dutch-process; both work well. The Dutch-process cake will be darker in color and slightly less sweet.
Try frosting this cake with Seven-Minute Frosting as shown here.
The butter in this recipe needs to be very soft. If you can leave a dent in the butter with a gentle touch, it's ready to use.
[Shares]
5.8k
Share on Facebook
Share by Email
Share on Pinterest
Share on Twitter",
  "notes": "Tips from our Bakers
The volume of this batter is 5 3/4 cups. It will rise quite a bit during baking. The weight of the batter is 3 pounds, 1 ounce. If you're weighing out layers, put 1 pound, 8 1/2 ounces in each 9\\" layer, or 1 pound in each 8\\" layer.
We've tested this cake with natural unsweetened cocoa, and with Dutch-process; both work well. The Dutch-process cake will be darker in color and slightly less sweet.
Try frosting this cake with Seven-Minute Frosting as shown here.
The butter in this recipe needs to be very soft. If you can leave a dent in the butter with a gentle touch, it's ready to use.",
  "source": "King Arthur Flour",
  "title": "Chocolate Cake",
  "totalTime": "",
  "yield": "",
}
`;
